src/Executor/Promise/Adapter/AmpPromiseAdapter.php

```diff
@@ -77,8 +79,13 @@ public function createRejected(Throwable $reason): Promise
         return new Promise($promise, $this);
     }
 
-    public function all(array $promisesOrValues): Promise
+    public function all(iterable $promisesOrValues): Promise
     {
+        assert(
+            is_array($promisesOrValues),
+            'AmpPromiseAdapter::all() requires arg to of array type'
+        );
```

Comment:
I think we can avoid this gotcha by replacing the `array_replace()` usage below with an implementation using `foreach`

Reply:
There's indexing into array below: `$promises[$key]` so key must be of `int|string`

Reply:
Cool, we can document that and allow `iterable<array-key, ?>`-

Reply:
Since it breaks covariance i'd rather not put it into method signature.

I'd keep arrays as they're in graphql-php implementations. We can eventually use better data structure in SyncPromise. for Amp and Reactphp we will have to wait for upstreams to accept iterable as well.

Reply:
See https://github.com/amphp/amp/blob/591e06ec427c2d2e5018172552b9354c208419e7/lib/functions.php#L387, we cannot help it much here now.

Reply:
When we violate covariance, we can provide at least some chance of catching an error statically in the callsite.

Reply:
For example, if `AmpPromiseAdapter::all()` were annotated to accept `iterable<array-key, ?>` and it was called with an `iterable<Foo, ?>`. Static analysis could catch such an error.

Reply:
I don't think so since `AmpPromiseAdapter::all()` is never called I guess, interface `PromiseAdapter::all()` is usually used. Therefore, `assert()` is more suitable here so at least tests can catch it. WDYT?

Reply:
Yeah right, it is not really called directly. The passed in value should always be in control of the developer and not change at runtime?

Reply:
Yup if I understand the question. I e.g. produce it through https://github.com/overblog/dataloader-php